File: node/json_file_tools/field_repetition.js
const fs = require('fs')
const path = require('path')

const target = process.argv[2]
const field = process.argv[3]
if (!target || !field) {
  console.error('usage: node field_repetition.js <folder_path or file_path> <field_name>')
  process.exit(1)
}
if (target.slice(-5).toLowerCase() === '.json') {
  processFile(target)
} else {
  processFolder(target)
}

// Process the given JSON file
function processFile(filePath) {
  console.log(`\n=== Processing file: ${filePath} ===`)
  const content = fs.readFileSync(filePath, 'utf-8')
  const regex = new RegExp(`"${field}"\\s*:\\s*("[^"]+")`, 'g')
  let match
  let values = []
  while ((match = regex.exec(content)) !== null) {
    values.push(match[1])
  }
  const duplicates = values.filter((value, index) => values.indexOf(value) !== index)
  if (duplicates.length === 0) {
    console.log(`No repetitions found.`) 
    return
  }
  if (duplicates.length > 0) {
    console.log(`  Repetitions found for field "${field}"`) 
    console.log('  Duplicates:')
    const uniqueDuplicates = [...new Set(duplicates)]
    uniqueDuplicates.forEach(dup => console.log(`  - ${dup}`))
  }
}

// Process all JSON files in the given folder
function processFolder(folderPath) {
  const files = getAllJsonFiles(folderPath)
  files.forEach(file => {
    processFile(file)
  })
}

// Recursively get all JSON files in the folder
function getAllJsonFiles(dirPath, fileList = []) {
  const entries = fs.readdirSync(dirPath)
  entries.forEach(entry => {
    const fullPath = path.join(dirPath, entry)
    if (fs.statSync(fullPath).isDirectory()) {
      getAllJsonFiles(fullPath, fileList)
    } else if (entry.slice(-5).toLowerCase() === '.json') {
      fileList.push(fullPath)
    }
  })
  return fileList
}

File: python/json_file_tools/field_repetition.py
import sys
import re

def main():
    if len(sys.argv) != 3:
        print("Usage: python field_repetition.py <folder_path or file_path> <field_name>")
        return
    else:
        target = sys.argv[1]
        field_name = sys.argv[2]
        if target.endswith('.json'):
            process_file(target, field_name)
        else:
            process_folder(target, field_name)


def process_file(file_path, field_name):
    print(f"\n=== Processing file: {file_path} ===")
    with open(file_path, 'r', encoding='utf-8') as f:
        content = f.read()
    regex = rf'"{field_name}"\s*:\s*"([^"]+)"'
    matches = re.findall(regex, content)
    seen = set()
    duplicates = set()
    for item in matches:
        if item in seen:
            duplicates.add(item)
        else:
            seen.add(item)
    if duplicates:
        print(f'  Repetitions found for field "{field_name}"')
        print('  Duplicates:')
        for dup in duplicates:
            print(f'    - {dup}')
    else:
        print('  No repetitions found.')

def process_folder(folder_path, field_name):
    import os
    for root, dirs, files in os.walk(folder_path):
        for file in files:
            if file.endswith('.json'):
                file_path = os.path.join(root, file)
                relative_path = os.path.join(folder_path, os.path.relpath(file_path, folder_path))
                process_file(relative_path, field_name)
            

        

if __name__ == "__main__":
    main()